pkg/apis/batch/validation/validation.go

```diff
@@ -486,9 +499,115 @@ func validatePodTemplateUpdate(spec, oldSpec batch.JobSpec, fldPath *field.Path,
+		}
+		isIndexed := ptr.Deref(job.Spec.CompletionMode, batch.NonIndexedCompletion) == batch.IndexedCompletion
+		if !isIndexed {
+			if job.Status.Succeeded < oldJob.Status.Succeeded {
+				allErrs = append(allErrs, field.Invalid(statusFld.Child("succeeded"), job.Status.Succeeded, "cannot decrease the succeeded counter"))
+			}
+		}
+		// In the checks below we fail validation for Job status fields (or conditions) only if they change their values
+		// (compared to the oldJob). This allows proceeding with status updates unrelated to the fields violating the
+		// checks, while blocking bad status updates for jobs with correct status.
+		if oldJob.Status.CompletionTime != nil {
+			if job.Status.CompletionTime == nil {
+				allErrs = append(allErrs, field.Required(statusFld.Child("completionTime"), "completionTime cannot be removed"))
+			} else {
+				allErrs = append(allErrs, apivalidation.ValidateImmutableField(job.Status.CompletionTime, oldJob.Status.CompletionTime, statusFld.Child("completionTime"))...)
+			}
+		}
+		if !ptr.Equal(oldJob.Status.CompletionTime, job.Status.CompletionTime) || isConditionTrue(oldJob.Status.Conditions, batch.JobComplete) != isConditionTrue(job.Status.Conditions, batch.JobComplete) {
+			if job.Status.CompletionTime != nil && !isConditionTrue(job.Status.Conditions, batch.JobComplete) {
+				allErrs = append(allErrs, field.Invalid(statusFld.Child("completionTime"), job.Status.CompletionTime, "cannot set completionTime when there is no Complete=True condition"))
+			}
+			if oldJob.Status.CompletionTime == nil && job.Status.CompletionTime == nil && isConditionTrue(job.Status.Conditions, batch.JobComplete) {
+				allErrs = append(allErrs, field.Required(statusFld.Child("completionTime"), "completionTime is required for Complete jobs"))
+			}
+		}
+		if oldJob.Status.StartTime != nil && !ptr.Deref(job.Spec.Suspend, false) {
+			if job.Status.StartTime == nil {
+				allErrs = append(allErrs, field.Required(statusFld.Child("startTime"), "startTime cannot be removed for unsuspended job"))
+			}
+		}
+		if !ptr.Equal(oldJob.Status.StartTime, job.Status.StartTime) || !ptr.Equal(oldJob.Status.CompletionTime, job.Status.CompletionTime) {
+			if job.Status.StartTime != nil && job.Status.CompletionTime != nil && job.Status.CompletionTime.Before(job.Status.StartTime) {
+				allErrs = append(allErrs, field.Invalid(statusFld.Child("completionTime"), job.Status.CompletionTime, "completionTime cannot be set before startTime"))
+			}
+		}
+		// validation when transitioning to terminal phase
+		for _, cType := range []batch.JobConditionType{batch.JobFailed, batch.JobComplete} {
+			if !isConditionTrue(oldJob.Status.Conditions, cType) && isConditionTrue(job.Status.Conditions, cType) {
+				if job.Status.StartTime == nil {
+					allErrs = append(allErrs, field.Required(statusFld.Child("startTime"), fmt.Sprintf("startTime is required for jobs with terminal %s=True condition", cType)))
+				}
+				if job.Status.Active > 0 {
```

Comment:
Aren't these a bit strong? Wouldn't it be better not to encode fields representing the state of the cluster (pods) for the job stage in the validation and focus mainly on the internal type consistency?

Reply:
Lots of discussion happened on a recent design about the meaning of `Complete`.  We agreed that `Complete` meant success+all-pods-gone. The all-pods-gone bit mattered for higher order constructs trying to decide if the resources being used by the job had been released.

If we keep that decision, then this is correct for Complete.

I didn't see a similarly thorough discussion of Failed, but it means a similar thing (I suspect this could be argued either way), then this would be similarly important.

Reply:
okay, let's add it if there is a consensus on this

Reply:
Yes, Failed and Complete means finished job - no running pods, and no further job reconciliations needed. So we want these fields to be 0 before marking the job as finished. This is what job controller does, but it was not expressed by validation before.